File: local-server/diff_utils.py
from dataclasses import dataclass

@dataclass
class Hunk:
    filepath: str
    text: str

@dataclass
class SearchReplaceChange:
    filepath: str
    search_block: str
    replace_block: str

def find_hunks(diff_string):
    hunks = []
    current_filename = ""
    current_lines = ""
    for line in diff_string.splitlines():
        if line.startswith("---"):
            continue
        elif line.lstrip().startswith("+++"):
            if len(current_filename) > 0:
                hunks.append(Hunk(current_filename, current_lines))
            current_filename = line[3:]
            current_lines = ""
        elif line.lstrip().startswith("@@"):
            if len(current_filename) > 0:
                hunks.append(Hunk(current_filename, current_lines))
            current_lines = ""
        else:
            current_lines += line
            current_lines += "\n"
    hunks.append(Hunk(current_filename, current_lines))
    return hunks

def parse_diff(diff_string):
    hunks = find_hunks(diff_string)
    search_replace_blocks = []

    for hunk in hunks:
        filepath = hunk.filepath
        text = hunk.text

        search_block = ""
        replace_block = ""

        for line in text.splitlines():
            if line.startswith("-"):
                search_block += " " + line[1:] + "\n"
            elif line.startswith("+"):
                replace_block += " " + line[1:] + "\n"
            else:
                search_block += line + "\n"
                replace_block += line + "\n"

        search_replace_blocks.append(
            SearchReplaceChange(filepath, search_block, replace_block)
        )
        filepath = ""
        search_block = ""
        replace_block = ""

    search_replace_blocks.append(
        SearchReplaceChange(filepath, search_block, replace_block)
    )

    return search_replace_blocks

from thefuzz import fuzz

@dataclass
class Match:
    block: str
    score: float

def line_relevant(line):
    return not(len(line.strip()) == 0 or line.startswith("#") or line.startswith("//"))

def find_best_match(query_code: str, original_code: str):
    query_code = query_code.strip()

    original_lines = original_code.splitlines()
    query_lines = query_code.splitlines()

    if len(query_lines) == 0:
        return Match("SUPERDOCSTHISSTRINGWILLNEVEREVERBEFOUND", 100)

    best_match = Match("", -1)

    for start_line in range(len(original_lines)):
        min_end = min(len(original_lines), max(start_line, start_line + len(query_lines) - 5)) # +/- 5 lines for tolerance
        max_end = min(len(original_lines), start_line + len(query_lines) + 5)
        for end_line in range(min_end, max_end):
            full_original_snippet = "\n".join(original_lines[start_line:end_line + 1])

            snippet_from_original = "\n".join([line for line in original_lines[start_line:end_line + 1] if line_relevant(line)]) # the loop already doesn't include max_end
            snippet_from_query = "\n".join([line for line in query_lines if line_relevant(line)])

            stripped_original = " ".join([line.strip() for line in snippet_from_original.splitlines()])
            stripped_query =  " ".join([line.strip() for line in snippet_from_query.splitlines()])

            score = fuzz.ratio(stripped_original, stripped_query)

            # Weighting the first and last lines by 3x
            score += 3*fuzz.ratio(original_lines[start_line], query_lines[0])
            score += 3*fuzz.ratio(original_lines[end_line], query_lines[-1])

            if score > best_match.score:
                best_match = Match(full_original_snippet, score)
    return best_match

File: extension/src/tools/diffutils.ts
import * as fuzz from 'fuzzball';
import * as Diff from 'diff';

class Hunk {
    filepath: string;
    text: string;

    constructor(filepath: string, text: string) {
        this.filepath = filepath;
        this.text = text;
    }
}

class SearchReplaceChange {
    filepath: string;
    searchBlock: string;
    replaceBlock: string;

    constructor(filepath: string, searchBlock: string, replaceBlock: string) {
        this.filepath = filepath;
        this.searchBlock = searchBlock;
        this.replaceBlock = replaceBlock;
    }
}

function findHunks(diffString: string) {
    const hunks = [];
    let currentFilename = "";
    let currentLines = "";
    const lines = diffString.split('\n');

    lines.forEach((line) => {
        if (line.startsWith('---')) {
            // skip
        } else if (line.startsWith('+++')) {
            if (currentFilename.length > 0) {
                hunks.push(new Hunk(currentFilename, currentLines));
            }
            currentFilename = line.slice(3);
            currentLines = "";
        } else if (line.startsWith('@@')) {
            if (currentFilename.length > 0) {
                hunks.push(new Hunk(currentFilename, currentLines));
            }
            currentLines = "";
        } else {
            currentLines += line + "\n";
        }
    });

    hunks.push(new Hunk(currentFilename, currentLines));
    return hunks;
}

export function parseDiff(diffString: string) {
    const hunks = findHunks(diffString);
    const searchReplaceBlocks: SearchReplaceChange[] = [];

    hunks.forEach((hunk) => {
        const filepath = hunk.filepath;
        const text = hunk.text;

        let searchBlock = "";
        let replaceBlock = "";

        const lines = text.split('\n');
        lines.forEach((line) => {
            if (line.startsWith('-')) {
                searchBlock += line.slice(1) + "\n";
            } else if (line.startsWith('+')) {
                replaceBlock += line.slice(1) + "\n";
            } else {
                searchBlock += line.slice(1) + "\n";
                replaceBlock += line.slice(1) + "\n";
            }
        });

        searchReplaceBlocks.push(new SearchReplaceChange(filepath, searchBlock, replaceBlock));
    });

    return searchReplaceBlocks;
}

class Match {
    block: string;
    score: number;

    constructor(block: string, score: number) {
        this.block = block;
        this.score = score;
    }
}

function lineRelevant(line: string) {
    const trimmed = line.trim();
    return !(trimmed.length === 0 || trimmed.startsWith('#') || trimmed.startsWith('//'));
}

export function findBestMatch(queryCode: string, originalCode: string) {
    queryCode = queryCode.trim();

    const originalLines = originalCode.split('\n');
    const queryLines = queryCode.split('\n');

    if (queryLines.length === 0) {
        return new Match("SUPERDOCSTHISSTRINGWILLNEVEREVERBEFOUND", 100);
    }

    let bestMatch = new Match("", -1);

    for (let startLine = 0; startLine < originalLines.length; startLine++) {
        const minEnd = Math.min(originalLines.length, Math.max(startLine, startLine + queryLines.length - 5));
        const maxEnd = Math.min(originalLines.length, startLine + queryLines.length + 5);

        for (let endLine = minEnd; endLine < maxEnd; endLine++) {
            const fullOriginalSnippet = originalLines.slice(startLine, endLine + 1).join('\n');

            const snippetFromOriginal = originalLines.slice(startLine, endLine + 1).filter(lineRelevant).join('\n');
            const snippetFromQuery = queryLines.filter(lineRelevant).join('\n');

            const strippedOriginal = snippetFromOriginal.split('\n').map(line => line.trim()).join(' ');
            const strippedQuery = snippetFromQuery.split('\n').map(line => line.trim()).join(' ');

            let score = fuzz.ratio(strippedOriginal, strippedQuery);

            score += 3 * fuzz.ratio(originalLines[startLine], queryLines[0]);
            score += 3 * fuzz.ratio(originalLines[endLine], queryLines[queryLines.length - 1]);

            if (score > bestMatch.score) {
                bestMatch = new Match(fullOriginalSnippet, score);
            }
        }
    }
    return bestMatch;
}

export function createSRBlocks(oldFile: string, newFile: string): string{
    let generatedDiff = Diff.createPatch("main.py",  oldFile, newFile);
    let searchReplaceChanges = parseDiff(generatedDiff);
    let fixedFile = oldFile;
    for(var i = 0; i < searchReplaceChanges.length; i++){
        if (searchReplaceChanges[i].searchBlock.length == 0) {
            continue;
        }

        searchReplaceChanges[i].searchBlock = findBestMatch(searchReplaceChanges[i].searchBlock, oldFile).block;
        fixedFile = fixedFile.replace(searchReplaceChanges[i].searchBlock, `<<<<<<< SEARCH
${searchReplaceChanges[i].searchBlock}
=======
${searchReplaceChanges[i].replaceBlock.trimEnd()}
>>>>>>> REPLACE`)
    }

    return fixedFile;
}